### PR TITLE
Fix for llvm 4.0.0 on centos

### PR DESCRIPTION
**Pull request by jcftang**
This addresses https://github.com/LLNL/spack/issues/3791

This change may impact macos, I have not tested on macos at all to verify this. However it does let me install the package on centos. This change was made according to http://lists.llvm.org/pipermail/llvm-dev/2014-February/070232.html

### CODE REVIEW

**Comment by davydden** [comment]
@jppelteret fix for the issue you had with `llvm`

**Comment by jppelteret** [comment]
Frustratingly this patch leads to me hitting #1616 on Linux Mint when building shared libraries. But its probably the right thing to do anyway.

```
../../../../../bin/llvm-tblgen: error while loading shared libraries: libLLVMSupport.so.4: cannot open shared object file: No such file or directory
```

[spack-build.out.txt](https://github.com/LLNL/spack/files/935189/spack-build.out.txt)

**Comment by adamjstewart** [comment]
@jppelteret So with this PR, `llvm+shared_libs` fails. Does `llvm~shared_libs` succeed? And on the latest develop branch, do either `llvm+shared_libs` or `llvm~shared_libs` work? Maybe we could only add this flag when building statically.

I'm going to do some tests on this locally.

**Comment by jcftang** [comment]
JTRF this PR does not build on macos, I'm in the middle of adding/testing on my mac to see if it builds.

**Comment by jppelteret** [comment]
@adamjstewart You're too quick :-) I'm just finished building without shared libs, working from commit 808e9b4. More specifically, with this variant
```
+all_targets+clang+compiler-rt~debug+gold+internal_unwind+libcxx~link_dylib~lldb~polly~shared_libs
```
This was successful :-)

Can I expect any major differences between the commit I've been working off of and the latest version of `develop`? Would you still like me to update and check again?

**Comment by adamjstewart** [comment]
> Can I expect any major differences between the commit I've been working off of and the latest version of develop?

I don't think anything significant has changed in the last 9 days.

**Comment by jcftang** [comment]
This is still building on my mac, but the change that I propose is to only allow this option for llvm 4.x or higher and on the linux platform.

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -389,6 +389,10 @@ def install(self, spec, prefix):
                 raise SpackException(
                     'The lldb variant requires the `+clang` variant')
 
+        platform = self.spec.architecture.platform
+        if spec.satisfies('@4.0.0:') and platform == 'linux':
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/llvm/package.py]: This should be equivalent to `spec.satisfies('@4.0.0: platform=linux')`.

Comment (jcftang) [inline var/spack/repos/builtin/packages/llvm/package.py]: The suggested change doesn't seem to work, I pushed up something that did work for me and is a bit cleaner than before.

**Comment by jcftang** [comment]
The code sign issues with lldb relates to this AFAIK ... https://github.com/Homebrew/legacy-homebrew/issues/39621

**Comment by jcftang** [comment]
Building llvm@4.0.0-lldb works on OSX with this PR.

**Comment by befritz** [comment]
was able to build llvm-4.0.0 on ubuntu 16.04 with this PR, was failing before.

**Comment by adamjstewart** [comment]
I tried testing this out on our CentOS 6 cluster, but unfortunately it fails with the following error message:
```
CMake Error at projects/libunwind/src/CMakeLists.txt:70 (message):
  Compiler doesn't support generation of unwind tables if exception support
  is disabled.  Building libunwind DSO with runtime dependency on C++ ABI
  library is not supported.
```
This occurs with or without your PR. Has anyone ever encountered this? I'm building LLVM with Intel if that makes a difference.

**Comment by jcftang** [comment]
I built llvm with gcc on centos7 and with clang on macos.

**Comment by adamjstewart** [comment]
So I still can't install `llvm` with `intel`, but that bug is unrelated to this PR. I can say that on CentOS 6, without this PR I run into the aforementioned bug. This PR fixes the bug for me. Thanks @jcftang!

Before I merge this, @jppelteret can you clarify whether this PR breaks `llvm+shared_libs` or whether it was already broken to begin with?

**Comment by jppelteret** [comment]
@adamjstewart Thanks for asking. This PR shifts the point where `llvm+shared_libs` fails to build on Linux Mint 17.3 (based on Ubuntu 14.04) from what is documented in #3791 at the end of the build to #1616 somewhere at the beginning. So I would say that this has not necessarily changes anything as it simply looks like there might (?) be an issue in building with shared libraries on all machines based off of Ubuntu 14.04. Although I was able to build that static version of the library, I wasn't able to link against it. But I'll be posting on the google group to ask about that (I don't think its related).

> was able to build llvm-4.0.0 on ubuntu 16.04 with this PR, was failing before

@befritz  May I ask, was this with or without `shared_libs` enabled?

**Comment by adamjstewart** [comment]
@befritz just started using Spack with me at ANL so I'm going to go out on a limb and say that he was almost certainly using whatever the default is (`~shared_libs`). I'll try `+shared_libs` out myself.

**Comment by befritz** [comment]
That is correct, I had not set anything for shared_libs.

**Comment by adamjstewart** [comment]
Ok, so without this PR, on CentOS 6, `spack install llvm+shared_libs` fails with the aforementioned error message. I'll try with this PR now to make sure it isn't introducing any new bugs.

**Comment by adamjstewart** [comment]
And with this PR, I was able to reproduce the other error message:
```
../../bin/llvm-tblgen: error while loading shared libraries: libLLVMSupport.so.4: cannot open shared object file: No such file or directory
```
In summary, neither `llvm@4.0.0+shared_libs` or `llvm@4.0.0~shared_libs` can be installed without this PR, as you run into:
```
file RPATH_CHANGE could not write new RPATH:
```
This PR fixes that problem (probably for both `llvm+shared_libs` and `llvm~shared_libs`) but introduces a new problem with `llvm+shared_libs`:
```
../../bin/llvm-tblgen: error while loading shared libraries: libLLVMSupport.so.4: cannot open shared object file: No such file or directory
```
Personally, I think we should merge this PR and try to figure out the new bug that was introduced in another PR. Does anyone disagree?

**Comment by adamjstewart** [comment]
@jcftang Is this still a WIP?

**Comment by jppelteret** [comment]
@adamjstewart So it looks like assessment of the state of compiling LLVM on Ubuntu-based OS's appears to align with mine. So I agree that this should be merged and the error with `+shared_libs` be dealt with in another PR (it looks like its not new anyway - #1616 - but I suspect that this package is typically compiled with `~shared_libs`, the default option).

**Comment by jcftang** [comment]
@adamjstewart I missed this earlier on before it was merged, but for the record, yes I agree with merging this and then figuring out what the new issue is later on.

**Pull request merged**